Update third party libraries page (#2058)
wpilibsuite/frc-docs: source/docs/software/vscode-overview/3rd-party-libraries.rst

3rd Party Libraries
===================

> Teams that are using non-PWM motor controllers or advanced sensors will most likely need to install external vendor dependencies. Please view the below section to see information on adding an external library.

Reviewer: ```suggestion
Teams that are using non-PWM motor controllers or advanced sensors will most likely need to install external vendor dependencies. 
```
"Please view the below section to see information on adding an external library." doesn't really add any information.

> What Are Vendor Dependencies?

Reviewer: I would restructure this a little. Libraries are how the vendors add their code. The vendor dependency is how the build system is told about the library.

-----------------------------

> A vendor dependency is a way for vendors like CTRE, and REV to add their own custom code that interfaces with their motor controllers and other devices. This way you can interact with your devices via CAN and have access to more complex and in-depth features than traditional PWM control.

Reviewer 2: ```suggestion
A vendor dependency is a way for vendors like CTRE, REV, and others to add their own custom code that interfaces with their motor controllers and other devices. This way you can interact with your devices via CAN and have access to more complex and in-depth features than traditional PWM control.
```

Reviewer 3: Avoid verbiage such as "like". Replace with "such-ass" or parenthesis. Additionally, avoid addressing the reader directly with "you". Instead refer to the reader as "the user" or "the team".

Managing Vendor Dependencies
----------------------------

> Vendor dependencies are installed on a per-project basis (so each robot project can have its own set of vendor dependencies). Vendor dependencies can be installed "online" or "offline". The offline installer will usually be provided by the vendor themselves if they have one and will usually come with additional debugging/utility programs.

Reviewer 3: ```suggestion
Vendor dependencies are installed on a per-project basis (so each robot project can have its own set of vendor dependencies). Vendor dependencies can be installed "online" or "offline". The "online" functionality is done by downloading the dependencies over the internet, while offline is typically provided by a vendor installer.
```

Additional debugging/utility programs is unnecessary information.

> Vendor dependencies are installed on a per-project basis (so each robot project can have its own set of vendor dependencies). Vendor dependencies can be installed "online" or "offline". The "online" functionality is done by downloading the dependencies over the internet, while offline is typically provided by a vendor-specific installer.

Reviewer: I think the recommendation down below to use offline installers might be better here.

> .. warning:: If you choose to install a vendor dependency via the "online" mode, you must reconnect your computer to the internet and rebuild about every 30 days otherwise the cache will clear, completely deleting your cached library install.

Reviewer 3: Refactor to avoid addressing the reader directly.

How Does It Work?

> ^^^^^^^^^^^^^^^^^

Reviewer 3: Don't use title case here, or title case the upper section title.

.. warning:: If you choose to install a vendor dependency via the "online" mode, you must reconnect your computer to the internet and rebuild about every 30 days otherwise the cache will clear, completely deleting your cached library install.

The JSON file allows specification of complex libraries with multiple components (Java, C++, JNI, etc.) and also helps handle some complexities related to simulation. Vendors choosing to provide a remote URL in the JSON also enable users to check for updates from within VS Code.
How Does It Work?
^^^^^^^^^^^^^^^^^

> .. note:: The vendor JSON files are actually processed by GradleRIO once they are in your projects ``vendordeps`` folder. If you are using another IDE, you will need to manually create a "vendordeps" folder in your project and copy any desired vendor JSON files from the "wpilib/YYYY" folder (where they should be placed by an offline installer) or download them directly from the vendor and place them into the folder in the project.

Reviewer 3: Why is this being removed?

Author: I figure it would probably be much better to get users to use Gradle for other IDEs (from the Command-Line section) especially since we mention using the command line. If we wanted to keep this note for "manual"ly adding vendordeps, we could probably put it into another section like we do for the command line / VS Code ways

> For LabVIEW teams, there might be a few new :guilabel:`Third Party` items on various palettes (specifically, one in :guilabel:`Actuators`, one in :guilabel:`Actuators` -> :guilabel:`Motor Control` labeled :guilabel:`CAN Motor`, and one in :guilabel:`Sensors`). These correspond to folders in ``C:\Program Files\National Instruments\LabVIEW YYYY\vi.lib\Rock Robotics\WPI\Third Party`` where ``YYYY`` is the current year - 1. If it's 2020, the directory would be ``LabVIEW 2019``.

Reviewer: The LabVIEW version information is no longer accurate, as LabVIEW 2020 is still being used in 2023. Since there hasn't been a consistent pastern since 2020, recommend just specifying 2020 rather then trying to specify a rule.

> For LabVIEW teams, you may notice a few new Third Party items on various palettes (specifically, one in :guilabel:`Actuators`, one in :guilabel:`Actuators` -> :guilabel:`Motor Control` labeled :guilabel:`CAN Motor`, and one in :guilabel:`Sensors`). These correspond to folders in ``C:\Program Files\National Instruments\LabVIEW YYYY\vi.lib\Rock Robotics\WPI\Third Party`` where ``YYYY`` is the current year - 1. If it's 2020, the directory would be ``LabVIEW 2019``.

Reviewer 3: Avoid addressing the reader directly and emphasize "Third Party" (guilabel or bold)

> In order to install third party libraries for LabVIEW, you will need to download the VIs from the vendor (typically via some sort of installer) and then you can drag and drop the third party VIs into the respective folder mentioned above as you would with any other VI.

Reviewer 3: This sentence is long and hard to read. Refactor into multiple sentences and avoid addressing the reader directly.

Select the desired libraries to add to the project by checking the box next to each, then click :guilabel:`OK`. The JSON file will be copied to the ``vendordeps`` folder in the project, adding the library as a dependency to the project.

If you want to install a vendor library in online mode, press :kbd:`Ctrl+Shift+P` and type WPILib or click on the WPILib icon in the top right to open the WPILib Command Palette and begin typing :guilabel:`Manage Vendor Libraries` and select it in the menu, and then click on :guilabel:`Install new libraries (online)` instead and copy + paste your vendor JSON URL.

> 

Reviewer 3: Avoid addressing the reader directly

Checking for Updates (Offline)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Since dependencies are version managed and done on a per-project bases, even if you have installed an updated library using an offline installer, you will need to :guilabel:`Manage Vendor Libraries` and select :guilabel:`Check for updates (offline)` for each project you wish to update.

Reviewer 3: I think it'd be better to break this into multiple sentences (2?)

Checking for Updates (Online)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Part of the JSON file that vendors may optionally populate is an online update location. If a library has an appropriate location specified, running :guilabel:`Check for updates (online)` will check if a newer version of the library is available from the remote location.

Removing a Library Dependency
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To remove a library dependency from a project, select :guilabel:`Manage Current Libraries` from the :guilabel:`Manage Vendor Libraries` menu, check the box for any libraries to uninstall and click :guilabel:`OK`. These libraries will be removed as dependencies from the project.

Command-Line
~~~~~~~~~~~~

Adding a vendor library dependency from the vendor URL can also be done through the command-line via a gradle task. Open a command-line instance at the project root, and enter ``gradlew vendordep --url=<url>`` where ``<url>`` is the vendor JSON URL. This will add the vendor library dependency JSON file to the ``vendordeps`` folder of the project. Vendor libraries can be updated the same way.

> The ``vendordep`` gradle task can also fetch vendordep JSONs from the user ``wpilib`` folder. To do so, pass ``FRCLOCAL/Filename.json`` as the file URL. For example, ``gradlew vendordep --url=FRCLOCAL/WPILibNewCommands.json`` will fetch the JSON for the new command-based framework.

Reviewer 3: ```suggestion
The ``vendordep`` gradle task can also fetch vendordep JSONs from the user ``wpilib`` folder. To do so, pass ``FRCLOCAL/Filename.json`` as the file URL. For example, ``gradlew vendordep --url=FRCLOCAL/WPILibNewCommands.json`` will fetch the JSON for the command-based framework.
```
